File: python/inspect_api/registry.py
from __future__ import annotations

import threading
import time
from collections.abc import Callable
from dataclasses import dataclass

from .audio_store import AudioStore
from .relay import InspectorRelay
from .types import SessionMeta

@dataclass
class _Entry:
    relay: InspectorRelay
    created_at: float
    model: str
    state_fn: Callable[[], str]
    audio_store: AudioStore | None = None

_REGISTRY: dict[str, _Entry] = {}
_LOCK = threading.Lock()

def register(
    session_id: str,
    relay: InspectorRelay,
    model: str,
    state_fn: Callable[[], str],
    audio_store: AudioStore | None = None,
) -> None:
    entry = _Entry(
        relay=relay,
        created_at=time.time(),
        model=str(model),
        state_fn=state_fn,
        audio_store=audio_store,
    )
    with _LOCK:
        _REGISTRY[session_id] = entry

def unregister(session_id: str) -> None:
    with _LOCK:
        entry = _REGISTRY.pop(session_id, None)
    if entry is not None:
        entry.relay.close()

def get_relay(session_id: str) -> InspectorRelay | None:
    with _LOCK:
        entry = _REGISTRY.get(session_id)
    return entry.relay if entry is not None else None

def get_audio_store(session_id: str) -> AudioStore | None:
    with _LOCK:
        entry = _REGISTRY.get(session_id)
    return entry.audio_store if entry is not None else None

def list_meta() -> list[SessionMeta]:
    with _LOCK:
        items = list(_REGISTRY.items())
    out: list[SessionMeta] = []
    for sid, e in items:
        try:
            state = e.state_fn()
        except Exception:
            state = "unknown"
        out.append(
            SessionMeta(
                id=sid,
                created_at=e.created_at,
                model=e.model,
                state=state,
                turn_count=e.relay.turn_count(),
                last_event_ts=e.relay.last_event_ts(),
            )
        )
    return out

def clear() -> None:
    with _LOCK:
        items = list(_REGISTRY.values())
        _REGISTRY.clear()
    for entry in items:
        entry.relay.close()

File: rust/src/inspect/registry.rs
#![allow(dead_code)]

use std::collections::HashMap;
use std::sync::{Arc, Mutex, OnceLock};
use std::time::SystemTime;

use super::relay::InspectorRelay;
use super::types::SessionMeta;

struct Entry {
    relay: Arc<InspectorRelay>,
    created_at: f64,
    model: String,
    state: Box<dyn Fn() -> String + Send + Sync>,
}

static REGISTRY: OnceLock<Mutex<HashMap<String, Entry>>> = OnceLock::new();

fn registry() -> &'static Mutex<HashMap<String, Entry>> {
    REGISTRY.get_or_init(|| Mutex::new(HashMap::new()))
}

pub fn register(
    session_id: &str,
    relay: Arc<InspectorRelay>,
    model: impl Into<String>,
    state_fn: impl Fn() -> String + Send + Sync + 'static,
) {
    let created_at = SystemTime::now()
        .duration_since(SystemTime::UNIX_EPOCH)
        .map(|d| d.as_secs_f64())
        .unwrap_or(0.0);
    let entry = Entry {
        relay,
        created_at,
        model: model.into(),
        state: Box::new(state_fn),
    };
    registry()
        .lock()
        .expect("inspect registry poisoned")
        .insert(session_id.to_string(), entry);
}

pub fn unregister(session_id: &str) {
    let entry = registry()
        .lock()
        .expect("inspect registry poisoned")
        .remove(session_id);
    if let Some(e) = entry {
        e.relay.close();
    }
}

pub fn get_relay(session_id: &str) -> Option<Arc<InspectorRelay>> {
    let g = registry().lock().expect("inspect registry poisoned");
    g.get(session_id).map(|e| e.relay.clone())
}

pub fn list_meta() -> Vec<SessionMeta> {
    let g = registry().lock().expect("inspect registry poisoned");
    g.iter()
        .map(|(id, e)| SessionMeta {
            id: id.clone(),
            created_at: e.created_at,
            model: e.model.clone(),
            state: (e.state)(),
            turn_count: e.relay.turn_count(),
            last_event_ts: e.relay.last_event_ts(),
        })
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::path::PathBuf;

    fn temp_dir() -> PathBuf {
        let p = std::env::temp_dir().join(format!(
            "speaches-plus-reg-{}-{}",
            std::process::id(),
            uuid::Uuid::new_v4().simple()
        ));
        std::fs::create_dir_all(&p).unwrap();
        p
    }

    #[test]
    fn register_and_lookup_round_trip() {
        let dir = temp_dir();
        let id = format!("sess_reg_{}", uuid::Uuid::new_v4().simple());
        let relay = Arc::new(InspectorRelay::new(id.clone(), Some(dir.clone())));
        register(&id, relay.clone(), "gpt-test", || "active".into());
        assert!(get_relay(&id).is_some());
        let metas = list_meta();
        assert!(metas.iter().any(|m| m.id == id && m.model == "gpt-test"));
        unregister(&id);
        assert!(get_relay(&id).is_none());
        let _ = std::fs::remove_dir_all(&dir);
    }
}
